### PR TITLE
Parse empirical dispersion corrections

### LINKED ISSUE #660: ORCA parser does not include dispersion correction in scfenergies
Hi,

Currently, the ORCA parser does not take into account dispersion corrections provided through post-SCF corrections, like the Grimme D3 correction.

I have attached an example below.

[wB97XD3_def2-SVP_toluene_S0_TDA6S_2-20.txt](https://github.com/cclib/cclib/files/2629257/wB97XD3_def2-SVP_toluene_S0_TDA6S_2-20.txt)

In this example, the pre-D3 SCF energy on the `Total Energy       :` line is `-306.98073026 Eh`, which is what cclib finds. However, the DFT+D3 energy is on the `FINAL SINGLE POINT ENERGY` line and is `-306.983847616167 Eh`.

I appreciate that this is possibly 'intended' behaviour, since strictly speaking, the D3 correction is post-SCF. However, this is inconsistent with, for example, Gaussian, since Gaussian includes the dispersion component in print outs of the SCF energy.

### LINKED ISSUE #664: Ensure dispersion correction is added to scfenergies for all packages
This is to help us keep track. It may not be relevant for all packages, but it should be investigated for each. Because this is relatively critical, the first pass should make sure `scfenergies` includes all relevant energies; for example, a 3-body energy may be optionally calculated, and might not be present in the final dispersion energy.

  * [ ] ADF (???)
  * [x] DALTON (???)
  * [x] Firefly (???)
  * [x] GAMESS-US (???)
~~  * [ ] GAMESS-UK (???)~~
  * [x] Gaussian (???)
~~  * [ ] Jaguar (???)~~
  * [x] Molcas (???)
  * [x] Molpro (???)
  * [x] MOPAC (???)
  * [x] NWChem (???)
  * [x] ORCA (issue: https://github.com/cclib/cclib/issues/660, PR: https://github.com/cclib/cclib/pull/418)
  * [x] Psi3 (???)
  * [x] Psi4 (???)
  * [x] Q-Chem (???)
  * [x] Turbomole (???)

### PR DESCRIPTION
**Pull request by berquist**
Extends #418 to all parsers. Closes #660. Closes #664.

### CODE REVIEW

**Comment by berquist** [comment]
Comment to myself: this is still draft while the following issues exist
```
********* SUMMARY OF ERRORS *********

basicPsi4-1.2.1/dvb_dispersion_bp86_d3zero.out: None
testdispersionenergies (test.data.testSP.GenericDispersionTest)Traceback (most recent call last):
  File "/home/eric/development/cclib_berquist/test/data/testSP.py", line 400, in testdispersionenergies
    self.assertTrue(len(self.data.dispersionenergies), 1)
  File "/home/eric/development/cclib_berquist/test/test_data.py", line 124, in ccdata_getattribute_with_coverage
    return object.__getattribute__(self, attr)
AttributeError: 'ccData' object has no attribute 'dispersionenergies'


********* SUMMARY OF FAILURES *********

basicFirefly8.1/dvb_dispersion_bp86_d3zero.out: None
testdispersionenergies (test.data.testSP.GenericDispersionTest)Traceback (most recent call last):
  File "/home/eric/development/cclib_berquist/test/data/testSP.py", line 401, in testdispersionenergies
    self.assertAlmostEquals(self.data.dispersionenergies[0], -0.40407836521053214)
AssertionError: -0.42998210046822655 != -0.40407836521053214 within 7 places (0.025903735257694405 difference)

basicGAMESS-US2018/dvb_dispersion_bp86_d3zero.out: None
testdispersionenergies (test.data.testSP.GenericDispersionTest)Traceback (most recent call last):
  File "/home/eric/development/cclib_berquist/test/data/testSP.py", line 401, in testdispersionenergies
    self.assertAlmostEquals(self.data.dispersionenergies[0], -0.40407836521053214)
AssertionError: -0.40054977293661714 != -0.40407836521053214 within 7 places (0.003528592273914999 difference)

basicGaussian16/dvb_dispersion_bp86_d3zero.out: None
testdispersionenergies (test.data.testSP.GenericDispersionTest)Traceback (most recent call last):
  File "/home/eric/development/cclib_berquist/test/data/testSP.py", line 401, in testdispersionenergies
    self.assertAlmostEquals(self.data.dispersionenergies[0], -0.40407836521053214)
AssertionError: -0.40054946000568914 != -0.40407836521053214 within 7 places (0.0035289052048430025 difference)

basicNWChem6.8/dvb_dispersion_bp86_d3zero.out: None
testdispersionenergies (test.data.testSP.GenericDispersionTest)Traceback (most recent call last):
  File "/home/eric/development/cclib_berquist/test/data/testSP.py", line 401, in testdispersionenergies
    self.assertAlmostEquals(self.data.dispersionenergies[0], -0.40407836521053214)
AssertionError: -0.4005497099422608 != -0.40407836521053214 within 7 places (0.0035286552682713412 difference)

basicORCA4.1/dvb_dispersion_bp86_d3zero.out: None
testdispersionenergies (test.data.testSP.GenericDispersionTest)Traceback (most recent call last):
  File "/home/eric/development/cclib_berquist/test/data/testSP.py", line 401, in testdispersionenergies
    self.assertAlmostEquals(self.data.dispersionenergies[0], -0.40407836521053214)
AssertionError: -0.4005497375618166 != -0.40407836521053214 within 7 places (0.003528627648715532 difference)

basicQChem5.1/dvb_dispersion_bp86_d3zero.out: None
testdispersionenergies (test.data.testSP.GenericDispersionTest)Traceback (most recent call last):
  File "/home/eric/development/cclib_berquist/test/data/testSP.py", line 401, in testdispersionenergies
    self.assertAlmostEquals(self.data.dispersionenergies[0], -0.40407836521053214)
AssertionError: -0.4005494953804897 != -0.40407836521053214 within 7 places (0.0035288698300424692 difference)
```

**Comment by berquist** [comment]
Comment to myself: this is still draft while the following issues exist
```
********* SUMMARY OF ERRORS *********

basicPsi4-1.2.1/dvb_dispersion_bp86_d3zero.out: None
testdispersionenergies (test.data.testSP.GenericDispersionTest)Traceback (most recent call last):
  File "/home/eric/development/cclib_berquist/test/data/testSP.py", line 400, in testdispersionenergies
    self.assertTrue(len(self.data.dispersionenergies), 1)
  File "/home/eric/development/cclib_berquist/test/test_data.py", line 124, in ccdata_getattribute_with_coverage
    return object.__getattribute__(self, attr)
AttributeError: 'ccData' object has no attribute 'dispersionenergies'


********* SUMMARY OF FAILURES *********

basicFirefly8.1/dvb_dispersion_bp86_d3zero.out: None
testdispersionenergies (test.data.testSP.GenericDispersionTest)Traceback (most recent call last):
  File "/home/eric/development/cclib_berquist/test/data/testSP.py", line 401, in testdispersionenergies
    self.assertAlmostEquals(self.data.dispersionenergies[0], -0.40407836521053214)
AssertionError: -0.42998210046822655 != -0.40407836521053214 within 7 places (0.025903735257694405 difference)

basicGAMESS-US2018/dvb_dispersion_bp86_d3zero.out: None
testdispersionenergies (test.data.testSP.GenericDispersionTest)Traceback (most recent call last):
  File "/home/eric/development/cclib_berquist/test/data/testSP.py", line 401, in testdispersionenergies
    self.assertAlmostEquals(self.data.dispersionenergies[0], -0.40407836521053214)
AssertionError: -0.40054977293661714 != -0.40407836521053214 within 7 places (0.003528592273914999 difference)

basicGaussian16/dvb_dispersion_bp86_d3zero.out: None
testdispersionenergies (test.data.testSP.GenericDispersionTest)Traceback (most recent call last):
  File "/home/eric/development/cclib_berquist/test/data/testSP.py", line 401, in testdispersionenergies
    self.assertAlmostEquals(self.data.dispersionenergies[0], -0.40407836521053214)
AssertionError: -0.40054946000568914 != -0.40407836521053214 within 7 places (0.0035289052048430025 difference)

basicNWChem6.8/dvb_dispersion_bp86_d3zero.out: None
testdispersionenergies (test.data.testSP.GenericDispersionTest)Traceback (most recent call last):
  File "/home/eric/development/cclib_berquist/test/data/testSP.py", line 401, in testdispersionenergies
    self.assertAlmostEquals(self.data.dispersionenergies[0], -0.40407836521053214)
AssertionError: -0.4005497099422608 != -0.40407836521053214 within 7 places (0.0035286552682713412 difference)

basicORCA4.1/dvb_dispersion_bp86_d3zero.out: None
testdispersionenergies (test.data.testSP.GenericDispersionTest)Traceback (most recent call last):
  File "/home/eric/development/cclib_berquist/test/data/testSP.py", line 401, in testdispersionenergies
    self.assertAlmostEquals(self.data.dispersionenergies[0], -0.40407836521053214)
AssertionError: -0.4005497375618166 != -0.40407836521053214 within 7 places (0.003528627648715532 difference)

basicQChem5.1/dvb_dispersion_bp86_d3zero.out: None
testdispersionenergies (test.data.testSP.GenericDispersionTest)Traceback (most recent call last):
  File "/home/eric/development/cclib_berquist/test/data/testSP.py", line 401, in testdispersionenergies
    self.assertAlmostEquals(self.data.dispersionenergies[0], -0.40407836521053214)
AssertionError: -0.4005494953804897 != -0.40407836521053214 within 7 places (0.0035288698300424692 difference)
```

**Comment by ghutchis** [comment]
Is this a difference between D3 "zero" and D3BJ?

**Comment by berquist** [comment]
@ghutchis if I remember correctly, it's *mostly* that, but there may have been something else. I started switching from D3(0) to D3(BJ) since not all programs I have access to can do D3(0), strangely. I'll check tonight, also to make sure I don't have changes in a stash.

**Comment by ghutchis** [comment]
The Firefly example seems totally out of left field.. You can always check with dftd3 to get the 'authentic' numbers: 
https://anaconda.org/psi4/dftd3

That's the Psi4 modified version, but IIRC Lori didn't change anything (https://github.com/loriab/dftd3)

**Comment by berquist** [comment]
Here's the current state of this PR. Some of these might just be stray untracked outputs in my work directories. All calculations using BP86.

| program  | version | D3 damping | energy                    |
|----------|---------|------------|---------------------------|
| DALTON   | 2018    | BJ         | -0.89908282               |
| DALTON   | 2018    | zero       | -0.40407837               |
| Firefly  | 8.1     | ?          | -0.4299821                |
| GAMESS   | 2018    | BJ         | -0.90884967               |
| GAMESS   | 2018    | zero       | -0.40054977               |
| GAMESS   | 2018    | ?          | -0.97852033               |
| Gaussian | 16      | zero       | -0.40054946               |
| Molpro   | 2018    | zero       | `mocoeffs` parsing broken |
| NWChem   | 6.8     | zero       | -0.40054971               |
| ORCA     | 4.0     | ?          | -4.24985755               |
| ORCA     | 4.1     | zero       | -0.40054974               |
| Psi4     | 1.2.1   | zero       | can't parse yet?          |
| Q-Chem   | 5.1     | zero       | -0.4005495                |

Filtered just for D3(0) and working,

| program  | version | energy      |
|----------|---------|-------------|
| DALTON   | 2018    | -0.40407837 |
| Firefly  | 8.1     | -0.4299821  |
| GAMESS   | 2018    | -0.40054977 |
| Gaussian | 16      | -0.40054946 |
| NWChem   | 6.8     | -0.40054971 |
| ORCA     | 4.1     | -0.40054974 |
| Q-Chem   | 5.1     | -0.4005495  |

It looks like the next reasonable tasks are:
- Get BP86/D3(0) corrections for the geometries in the DALTON and Firefly inputs.
  - If `dftd3` is consistent with DALTON and Firefly, then the geometries are different; update the geometries and rerun. Consistent geometries is already a known issue: https://github.com/cclib/cclib/issues/407
  - If `dftd3` is consistent with every other package, then there is something up with the Firefly and DALTON settings or implementations; make specific unit tests and link to this comment.
- Run the ORCA 4.2 calculation.
- Implement for Psi4.

Anything for Molpro will have to wait until https://github.com/cclib/cclib/pull/872/ is done, which will take more time.

**Comment by berquist** [comment]
The Q-Chem and Firefly results with `dftd3` are identical, but comparing against the DALTON geometry:

| program's input geometry | energy (a.u.) | energy (eV) using cclib converter |
|--------------------------|---------------|-----------------------------------|
| Q-Chem                   | -0.01471993   | -0.40054968313904654              |
| DALTON                   | -0.01484961   | -0.40407845555233046              |

This indicates that
- the geometries we use with DALTON are inconsistent, and
- something else is going on with Firefly.

**Comment by berquist** [comment]
I've updated the geometry for the DALTON calculation and now the results are identical. To take a look at the raw outputs, we can see that at least part of the Firefly parameters are different from DALTON and Q-Chem:

| param | DALTON | Q-Chem | Firefly |
|-------|--------|--------|---------|
| s6    | 1.000  | 1.000  | 1.000   |
| s8    | 1.683  | 1.683  | 1.682   |
| sr6   | 1.139  | 1.139  | 1.094   |
| sr8   | 1.000  | 1.000  | 1.000   |
| alpha | 14.000 | ?      | ?       |
| k1    | ?      | ?      | 16      |
| k2    | ?      | ?      | 1.333   |
| k3    | ?      | ?      | -4      |
| rthr  |        |        | 20000   |
| rthr2 | ?      | ?      | 1600    |

The nuclear repulsion energy is also wrong, even before adding the dispersion correction.

DALTON:
```
                      DFT-D Empirical Dispersion Correction
                      -------------------------------------

       Running DFT-D Version :     3
                   Reference : S. Grimme, J. Antony, S. Ehrlich and H. Krieg, 
                               J. Chem. Phys. 132 (2010), 154104
       Becke-Johnson Damping :    F
                3-body terms :    F


                      DFT-D3 Functional Dependent Parameters
                      --------------------------------------

      s_6 =       1.000000000000     sr_6 =       1.139000000000
      s_8 =       1.683000000000     sr_8 =       1.000000000000
      alp =      14.000000000000


                          +------------------------------+
                          ! Dispersion Energy Correction !
                          +------------------------------+

                      E_disp :         -0.014719923805
```
Q-Chem:
```
 Adding empirical dispersion correction 
 ----------------------------------------------------------------
 Doing -D3(0) with
 s6  =   1.00000
 s8  =   1.68300
 rs6 =   1.13900
 rs8 =   1.00000
 -D3 energy without 3body term =     -0.0147199231 hartrees
```
Firefly:
```
Loading Firefly/DFT-D extension... loaded successfully (version  1.0001).

 Firefly/DFT-D extension capabilities:
 Minimum supported DFT-D version is   2
 Maximum supported DFT-D version is   4
 For DFT-D version   2, max supported atomic no. is  86
 For DFT-D version   3, max supported atomic no. is  94
 For DFT-D version   4, max supported atomic no. is  94

 Using DFT-D version  3
 Cite DFT-D3 as:
 S. Grimme, J. Antony, S. Ehrlich and H. Krieg,
 J. Chem. Phys. 132 (2010), 154104

 Initializing DFT-D extension... initialized successfully!

     ---------------------------------------------
         Parameters controlling DFT-D run are:
     ---------------------------------------------
     VERSN =             3   S6    = 1.0000000E+00
     RS6   = 1.0940000E+00   S18   = 1.6820000E+00
     RS18  = 1.0000000E+00   ALP   = 1.4000000E+01
     K1    = 1.6000000E+01   K2    = 1.3333333E+00
     K3    =-4.0000000E+00   RTHR  = 2.0000000E+04
     RTHR2 = 1.6000000E+03   NUMGRD=             F
     ABC   =             F   TZ    =             F
```

**Comment by berquist** [comment]
Molpro is implemented, but not tested, since there are bigger problems with parsing MO coefficients from the latest version (https://github.com/cclib/cclib/pull/872).

**Comment by shivupa** [comment]
Sorry for the delay. This looks good to me, and I'm in favor of merging. I just have a question:

>The nuclear repulsion energy is also wrong, even before adding the dispersion correction.
The firefly nuclear repulsion energy? Can this and the difference in parameters be reported upstream?

File: cclib/parser/orcaparser.py [hunk 1]
```diff
@@ -293,6 +323,47 @@ def splitter(line):
...
+-------------------------   ----------------
+Dispersion correction           -0.016199959
+-------------------------   ----------------
+"""
+        if "DFT DISPERSION CORRECTION" in line:
+            # A bunch of parameters are printed the first time dispersion is called
+            # However, they vary wildly in form and number, making parsing problematic
+            line = next(inputfile)
+            while 'Dispersion correction' not in line:
+                line = next(inputfile)
```
Comment (shivupa) [inline cclib/parser/orcaparser.py:363]: Should we do `self.skip_lines(inputfile,["Dispersion parameters"])``

Comment (berquist) [reply cclib/parser/orcaparser.py:363]: It isn't clear if the number of lines to skip is fixed, and we (I?) do this in other places.

**Comment by berquist** [comment]
> Can this and the difference in parameters be reported upstream?

I see in the 8.2.0 (latest version) changelog

> 8. Changes to use newer, more precise value of Bohr radius, namely 
   0.52917721067 Angstrom, throughout all the code. Previous versions 
   of Firefly used older, less precise value of 0.52917724924 Angstrom.
   Note: this transition causes minor differences in computed energies and 
   other properties as compared to older Firefly versions when initial 
   geometry is specified using Angstroms.

which may account for this (I used 8.1.1). It's possible that 8.2.0 has fixed this. I have the 8.2.0 archive, but never registered it to get the password.

**Comment by berquist** [comment]
> I think there's some details to fill in here.

I can't think of anything off the top of my head. What do you think is missing?

**Pull request merged**